Mathlib/RingTheory/AdicCompletion/Algebra.lean

```diff
@@ -59,11 +64,35 @@ def subalgebra : Subalgebra R (∀ n, R ⧸ (I ^ n • ⊤ : Ideal R)) :=
 def subring : Subring (∀ n, R ⧸ (I ^ n • ⊤ : Ideal R)) :=
   Subalgebra.toSubring (subalgebra I)
 
-instance : CommRing (AdicCompletion I R) :=
-  inferInstanceAs <| CommRing (subring I)
-
-instance : Algebra R (AdicCompletion I R) :=
-  inferInstanceAs <| Algebra R (subalgebra I)
+@[irreducible]
+instance : Mul (AdicCompletion I R) where
+  mul x y := ⟨x.val * y.val, by simp [x.property, y.property]⟩
+
+@[irreducible]
+instance : One (AdicCompletion I R) where
+  one := ⟨1, by simp⟩
+
+instance : CommRing (AdicCompletion I R) where
+  left_distrib x y z := Subtype.ext <| left_distrib x.val y.val z.val
+  right_distrib x y z := Subtype.ext <| right_distrib x.val y.val z.val
+  zero_mul x := Subtype.ext <| zero_mul x.val
+  mul_zero x := Subtype.ext <| mul_zero x.val
+  mul_assoc x y z := Subtype.ext <| mul_assoc x.val y.val z.val
+  one_mul x := Subtype.ext <| one_mul x.val
+  mul_one x := Subtype.ext <| mul_one x.val
+  zsmul := zsmulRec
+  add_left_neg x := Subtype.ext <| add_left_neg x.val
+  sub_eq_add_neg x y := Subtype.ext <| sub_eq_add_neg x.val y.val
+  mul_comm x y := Subtype.ext <| mul_comm x.val y.val
+
+instance : Algebra R (AdicCompletion I R) where
+  toFun r := ⟨algebraMap R (∀ n, R ⧸ (I ^ n • ⊤ : Ideal R)) r, by simp⟩
```

Comment:
Lean might try to unfold `toFun` unnecessarily during unification checks. You could try hoisting it out as a separate declaration.